### PR TITLE
CDK: provide better user-friendly error messages

### LINKED ISSUE #12537: CDK: Connectors can define a method to get error message
## Tell us about the problem you're trying to solve
After #12472, connectors using the CDK will generically emit `AirbyteTraceMessage`s when encountering unhandled exceptions. These `AirbyteTraceMessage`s contain a generic, uninformative `message` that is displayed to the user.

The goal is to give developers an easier way to provide a better user-friendly messages based on raised exceptions.

## Describe the solution you’d like
Provide an overridable `get_error_display_message` (name TBD) method at the `Source` or `Stream` level that is called when an exceptions occurs while reading. The method receives a param with the exception that was raised so that the developer can then return a user-friendly message if appropriate.

(rough example)
```python
def get_error_display_message(error):
    if issubclass(error, MyErrorType):
        return "The stream xyz is disabled for this acccount"
    return None
```

(optional)
We could then implement this in our own HTTPStreams to extract error messages from API responses when `raise_on_http_errors` is enabled, either directly or by exposing a new method where the developer can use the response itself to return the message 

(rough example)
```python
def parse_error_response(response):
    return response.json()["error"]["message"]
```

This approach can fit in nicely with connectors-as-config later down the line by allowing devs to specify the json path to the error message.

TODO:
- check some existing connector apis and see how the errors are returned 

## Describe the alternative you’ve considered or used
Connector developers can currently choose to emit `AirbyteTraceMessage`s themselves with a user-friendly `message`, but this means catching exceptions themselves and building up / being aware of `AirbyteTaceMessage`.

### PR DESCRIPTION
**Pull request by pedroslopez**
## What

- Adds a way for developers to provide user-friendly display messages for errors when reading, to be emitted as `AirbyteTraceMessage`s.
- Uses the above to implement default error message retrieval logic for API errors coming from HTTPStreams.

closes #12537 

## How
- `Stream` gains a `get_error_display_message()` method, that takes in the exception and should return a `string` with a user-friendly error message for the exception (or `None`)
- `HTTPStream` implements `get_error_display_message` by handling `HTTPError`s and passing it to a new `parse_response_error_message` method.
- `HTTPStream.parse_error_message` contains logic to pull an error message from common API patterns. Developers can override this method if they want to.

## Recommended reading order
1. `airbyte_cdk/sources/abstract_source.py` - error handling actually happens here
2. `airbyte_cdk/sources/streams/core.py` - defines the `get_error_display_message` on all Streams
3. `airbyte_cdk/sources/streams/http/http.py` - overrides the default `get_error_display_message` on HTTPStreams to handle HTTPErrors. Defines `parse_response_error_message` method to allow overriding parsing logic.
4. `unit_tests/sources/streams/http/test_http.py`-  has examples of some responses that are handled

## 🚨 User Impact 🚨
Users will start to see error messages as provided by the APIs on HTTPStreams automatically if the API adheres to common practices. Connector developers can tweak this behavior as needed. 

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

## Tests

<details><summary><strong>Unit</strong></summary>

```

Results (4.00s):
     410 passed

```

</details>

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py [hunk 1]
```diff
@@ -336,6 +336,37 @@ def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mappi
             backoff_handler = default_backoff_handler(max_tries=max_tries, factor=self.retry_factor)
             return backoff_handler(user_backoff_handler)(request, request_kwargs)
 
+    def parse_response_error_message(self, response: requests.Response) -> Optional[str]:
+        # default logic to grab error from common fields
+
+        def _try_get_error(value):
+            if type(value) is str:
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py:343]: when checking if something is a particular type in python it's better to use `isinstance` since it considers inheritance

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py [hunk 2]
```diff
@@ -336,6 +336,37 @@ def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mappi
...
+                    or value.get("error")
+                    or value.get("errors")
+                    or value.get("failures")
+                    or value.get("failure")
+                )
+                return _try_get_error(new_value)
+            return None
+
+        try:
+            body = response.json()
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py:360]: we shouldn't assume the response is JSON

File: airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py [hunk 3]
```diff
@@ -3,13 +3,15 @@
...
 
-from airbyte_cdk.sources import Source
 from airbyte_cdk.utils.mapping_utils import all_key_pairs_dot_notation, get_value_by_dot_notation
 
+if TYPE_CHECKING:
+    from airbyte_cdk.sources import Source
 
-def get_secrets(source: Source, config: Mapping[str, Any], logger: logging.Logger) -> List[Any]:
+
+def get_secrets(source: "Source", config: Mapping[str, Any], logger: logging.Logger) -> List[Any]:
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py:14]: jc why was this needed?

**Review by sherifnada — COMMENTED** [review]
@pedroslopez the approach makes sense to me!

File: airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py (hunk 3 above)
Comment (pedroslopez) [reply airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py:14]: The usage of `Source` here for typing caused a circular dependency (this is used by AirbyteTraceMessage). Could also move this `get_secrets` method out into another file since AirbyteTraceMessage doesn't need this specific method (it uses other things defined in this file)

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py (hunk 2 above)
Comment (pedroslopez) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py:366]: Yeah, at the moment I'm returning `None` if a `JSONDecodeError` is encountered... Breaking it out into a separate method for decoding feels like it may cause confusion (this would only be used for parsing error responses but parsing regular responses is left up to you?). I think in this case you would override the method as needed to fit your use case.

File: airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py [hunk 4]
```diff
@@ -442,3 +442,48 @@ def test_send_raise_on_http_errors_logs(mocker, status_code):
...
+        ({"errors": ["one", "two", "three"]}, "one, two, three"),
+        ({"messages": ["one", "two", "three"]}, "one, two, three"),
+        ({"errors": [{"message": "one"}, {"message": "two"}, {"message": "three"}]}, "one, two, three"),
+        ({"errors": [{"error": "one"}, {"error": "two"}, {"error": "three"}]}, "one, two, three"),
+        (["one", "two", "three"], "one, two, three"),
+        ({"error": True}, None),
+        ({"something_else": "hi"}, None),
+        ({}, None),
+    ],
+)
```
Comment (evantahler) [inline airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py:462]: I love this test-case based development

Comment (evantahler) [reply airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py:462]: I think a missing case would be:

```
({"error": {"code": "foo-123"}}, "foo-123"),
```

Showing that we prefer to show /something/ from an error object.

What about synonyms for "error"?

```
({"failure": {"message": "foo-123"}}, "foo-123"),
```

File: airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py [hunk 5]
```diff
@@ -442,3 +442,48 @@ def test_send_raise_on_http_errors_logs(mocker, status_code):
...
         stream.logger.error.assert_called_with(response.text)
         assert response.status_code == status_code
+
+
+@pytest.mark.parametrize(
+    "api_response, expected_message",
+    [
+        ({"error": "something broke"}, "something broke"),
+        ({"error": {"message": "something broke"}}, "something broke"),
+        ({"error": "err-001", "message": "something broke"}, "something broke"),
```
[the remark this answers is not in the document]
Comment (pedroslopez) [reply airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py:452]: This one I got from brian's comment on #12537:

```
{
    "error": "auth-0001",
    "message": "Incorrect username and password",
    "detail": "Ensure that the username and password included in the request are correct",
    "help": "https://example.com/help/error/auth-0001"
}
```

Where I believe the "user-friendly error message" would be the `message` key. They're theoretical examples though, so I could go either way...

**Review by sherifnada — APPROVED** [review]

File: airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (evantahler) [reply airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py:452]: I guess I agree with that, assuming there isn't an error object with additional fields. I don't feel too strongly either way - we'll be revising this a lot, I'm sure.

**Review by evantahler — APPROVED** [review]
Approved to move this forward, but I still think some additional error-cases / synonyms would be nice

File: airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py (hunk 4 above)
Comment (pedroslopez) [reply airbyte-cdk/python/unit_tests/sources/streams/http/test_http.py]: Added some more cases in [28c355d](https://github.com/airbytehq/airbyte/pull/12944/commits/28c355d5e5080f3c6009cb5853194b54cc777b2e)! I left out `code` for now since I'm not sure how useful it will be for end users to see. We can always add it later though if it does end up being useful.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py [hunk 6]
```diff
@@ -297,7 +296,7 @@ def _send(self, request: requests.PreparedRequest, request_kwargs: Mapping[str,
             # Raise any HTTP exceptions that happened in case there were unexpected ones
             try:
                 response.raise_for_status()
-            except HTTPError as exc:
+            except requests.HTTPError as exc:
```
[the remark this answers is not in the document]
Comment (marcosmarxm) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py:299]: @pedroslopez this should not be from `requests.exceptions.HTTPError`?
See one example of this throws the error:
```
2022-05-26 04:38:32 e[44msourcee[0m > Syncing stream: feedback_submissions 
2022-05-26 04:38:32 e[44msourcee[0m > Encountered an exception while reading stream SourceHubspot
Traceback (most recent call last):
  File "/airbyte/integration_code/source_hubspot/streams.py", line 340, in read_records
    stream_records, response = self._read_stream_records(
  File "/airbyte/integration_code/source_hubspot/streams.py", line 311, in _read_stream_records
    response = self.handle_request(
  File "/airbyte/integration_code/source_hubspot/streams.py", line 288, in handle_request
    response = self._send_request(request, request_kwargs)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py", line 337, in _send_request
    return backoff_handler(user_backoff_handler)(request, request_kwargs)
  File "/usr/local/lib/python3.9/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py", line 299, in _send
    response.raise_for_status()
  File "/usr/local/lib/python3.9/site-packages/requests/models.py", line 953, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://api.hubapi.com/crm/v3/objects/feedback_submissions?hapikey=*****&limit=100&startTimestamp=1485302400000&endTimestamp=1653539911984&archived=false&associations=contacts&properties=hs_agent_email%2Chs_agent_id%2Chs_agent_name%2Chs_all_accessible_team_ids%2Chs_all_assigned_business_unit_ids%2Chs_all_owner_ids%2Chs_all_team_ids%2Chs_chatflow_name%2Chs_chatflow_object_id%2Chs_contact_id%2Chs_conversation_thread_id%2Chs_created_by_user_id%2Chs_createdate%2Chs_lastmodifieddate%2Chs_merged_object_ids%2Chs_object_id%2Chs_unique_creation_key%2Chs_updated_by_user_id%2Chs_user_ids_of_all_notification_followers%2Chs_user_ids_of_all_notification_unfollowers%2Chs_user_ids_of_all_owners%2Chubspot_owner_assigneddate%2Chubspot_owner_id%2Chubspot_team_id%2Chs_industry_standard_question_type%2Chs_sentiment%2Chs_survey_id%2Chs_survey_type%2Chs_survey_channel%2Chs_submission_timestamp%2Chs_value%2Chs_response_group%2Chs_content%2Chs_ingestion_id%2Chs_knowledge_article_id%2Chs_visitor_id%2Chs_engagement_id%2Chs_submission_url%2Chs_survey_name%2Chs_form_guid%2Chs_contact_email_rollup%2Chs_submission_name

```

Looks is not capturing the correct Error Exception and showing the response from the API.
